### PR TITLE
fixed reading test adapter paths from runsettings

### PR DESCRIPTION
**Pull request by rahulpaw**
Fix for Issue #236.
Reads the test adapter path from runsettings.

Testing:
Added UT, Enabled Acceptance test.
validated console and IDE scenario.
Ran Test.cmd

### CODE REVIEW

File: src/Microsoft.TestPlatform.Common/Utilities/RunSettingsUtilities.cs [hunk 1]
```diff
@@ -112,5 +113,31 @@ public static int GetMaxCpuCount(RunConfiguration runConfiguration)
...
+        /// Gets the test adapters path from the run configuration
+        /// </summary>
+        /// <param name="runSettings">Test run settings</param>
+        /// <param name="returnNullIfNotSet">True to return null, if adapter paths is not set.</param>
+        /// <returns>Test adapters paths</returns>
+        public static IEnumerable<string> GetTestAdaptersPaths(string runSettings)
+        {
+            var runConfiguration = XmlRunSettingsUtilities.GetRunConfigurationNode(runSettings);
+
+            IEnumerable<string> testAdaptersPaths = null;
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Utilities/RunSettingsUtilities.cs]: Please return Enumerable.Empty instead. The consumer can directly do a foreach.

File: src/Microsoft.TestPlatform.Common/Utilities/RunSettingsUtilities.cs [hunk 2]
```diff
@@ -112,5 +113,31 @@ public static int GetMaxCpuCount(RunConfiguration runConfiguration)
...
+            if (runConfiguration != null)
+            {
+                if (!runConfiguration.TestAdaptersPathsSet)
```
Comment (codito) [inline src/Microsoft.TestPlatform.Common/Utilities/RunSettingsUtilities.cs]: We can invert `if`?

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 3]
```diff
@@ -121,5 +128,34 @@ public void UpdateExtensions(IEnumerable<string> pathToAdditionalExtensions, boo
...
+        private void UpdateTestAdapterPaths(string runSettings)
+        {
+            IEnumerable<string> customTestAdaptersPaths = RunSettingsUtilities.GetTestAdaptersPaths(runSettings);
+
+            if (customTestAdaptersPaths != null)
+            {
+                foreach (string customTestAdaptersPath in customTestAdaptersPaths)
+                {
+                    string adapterPath = string.Empty;
+                    adapterPath = Path.GetFullPath(Environment.ExpandEnvironmentVariables(customTestAdaptersPath));
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: Merge this and above line to `var adapterPath`

File: test/Microsoft.TestPlatform.Common.UnitTests/Utilities/RunSettingsUtilitiesTests.cs [hunk 4]
```diff
@@ -93,6 +95,29 @@ public void GetMaxCpuCountWithInvalidCpuCountShouldReturnDefaultCpuCount()
...
         }
+
+        [TestMethod]
+        public void GetTestAdaptersPaths()
+        {
+            string settingXml = @"<RunSettings><RunConfiguration><TestAdaptersPaths>C:\testadapterpath;D:\secondtestadapterpath</TestAdaptersPaths></RunConfiguration ></RunSettings>";
+            string[] expectedResult = new string[] { @"C:\testadapterpath", @"D:\secondtestadapterpath" };
+
+            string[] result = (string[])RunSettingsUtilities.GetTestAdaptersPaths(settingXml);
+            if (expectedResult.Length == result.Length)
```
Comment (codito) [inline test/Microsoft.TestPlatform.Common.UnitTests/Utilities/RunSettingsUtilitiesTests.cs]: Please use `CollectionAssert` methods to compare results.

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 5]
```diff
@@ -121,5 +128,34 @@ public void UpdateExtensions(IEnumerable<string> pathToAdditionalExtensions, boo
...
+                    if (!Directory.Exists(adapterPath))
+                    {
+                        EqtTrace.Warning(string.Format("AdapterPath Not Found:", adapterPath));
+                        continue;
+                    }
+                    List<string> adapterFiles = new List<string>(
```
Comment (codito) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: This logic is duplicated across many files. Should we consolidate?

**Comment by codito** [comment]
@rahulpaw with this change, will the final testadapterpaths continue to be a union of
0. Path specified in `-- configuration.testadapterpaths` in command line
1. Path specified in `/testAdapterPath` command line argument
2. Path specified in `<TestAdapterPaths>` in runsettings

File: src/Microsoft.TestPlatform.Common/Utilities/RunSettingsUtilities.cs (hunk 1 above)
Comment (rahulpaw) [inline src/Microsoft.TestPlatform.Common/Utilities/RunSettingsUtilities.cs]: addressed

**Review by codito — APPROVED** [review]

File: src/Microsoft.TestPlatform.Client/TestPlatform.cs [hunk 6]
```diff
@@ -76,6 +82,8 @@ public ITestRunRequest CreateTestRunRequest(TestRunCriteria testRunCriteria)
                 throw new ArgumentNullException("testRunCriteria");
             }
 
+            UpdateTestAdapterPaths(testRunCriteria.TestRunSettings);
+
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: This caused issue in Tpv1, when in VS IDE scenario data collectors were enabled, & runsettings were used to provide adapter path, it resulted in DataCollection(fakes, CodeCoverage) not working.

The workaround we used was to not update them here, but instead like we process TestAdapterPath in CLI, similarly process it, & apply it on ITestPlatform when provided via runsettings

Comment (rahulpaw) [inline src/Microsoft.TestPlatform.Client/TestPlatform.cs]: I would like to take a deeper look at the datacollector scenarios. but right now TPv2 does not support DataCollectors.

**Pull request merged**